pkg/service/storage/ucan.go

```diff
@@ -41,41 +41,72 @@ func NewUCANServer(storageService Service) (server.ServerView, error) {
 					if cap.With() != iCtx.ID().DID().String() {
 						return blob.AllocateOk{}, nil, NewUnsupportedCapabilityError(cap)
 					}
-
 					// check if we already have an allcoation for the blob in this space
 					allocs, err := storageService.Blobs().Allocations().List(ctx, digest)
 					if err != nil {
 						log.Errorf("getting allocations: %w", err)
 						return blob.AllocateOk{}, nil, failure.FromError(err)
 					}
 
+					allocated := false
 					for _, a := range allocs {
-						// if we find an allocation, check if we have the blob.
 						if a.Space == cap.Nb().Space {
-							_, err := storageService.Blobs().Store().Get(ctx, digest)
-							if err == nil {
-								// if we have it, it does not need upload
-								return blob.AllocateOk{Size: 0}, nil, nil
-							}
-							if !errors.Is(err, store.ErrNotFound) {
-								log.Errorf("getting blob: %w", err)
-								return blob.AllocateOk{}, nil, failure.FromError(err)
-							}
+							allocated = true
+							break
+						}
+					}
+
+					received := false
+					// check if we received the blob (only possible if we have an allocation)
+					if len(allocs) > 0 {
+						_, err = storageService.Blobs().Store().Get(ctx, digest)
+						if err == nil {
+							received = true
+						}
+						if err != nil && !errors.Is(err, store.ErrNotFound) {
+							log.Errorf("getting blob: %w", err)
+							return blob.AllocateOk{}, nil, failure.FromError(err)
 						}
 					}
 
+					// the size reported in the receipt is the number of bytes allocated
+					// in the space - if a previous allocation already exists, this has
+					// already been done, so the allocation size is 0
```

Comment:
q: does that imply that only the first space where a given blob was allocated is billed for that storage or this has nothing to do with charging users?

Reply:
Yes originally this was for billing purposes. In the case of failed uploads the user may try to upload the same data again, or sometimes they just uplaod the same thing multiple times. However we don't want to change them multiple times so the `size` value accounts for this. We need to revisit what this means in a decentralized network, as this particular node does not have a view over the whole network and there is no guarantee a second allocation will come to the same node as before (would be useful though).

Reply:
thanks for the context. I thought we were looking for previous allocations of the same blob in other spaces too, but I see now the check is scoped only to the space the invocation refers to.

about the issue in the decentralized scenario, would it be a possibility to store the DIDs of the spaces where a CID lives as part of the metadata when indexing? Storage nodes could then check with the indexing-service whether a given blob has already been allocated in that same space. I'm still missing a lot of context, so please let me know if that's just nonsensical 😄.

Reply:
Yes sounds like a reasonable suggestion!

Reply:
I've opened an issue with more details. https://github.com/storacha/storage/issues/23 - would you mind adding your thoughts?